File: sdks/typescript/src/monitor.ts
/**
 * Background File Monitoring for AvocadoDB
 *
 * Framework-agnostic file watcher that detects changes and triggers re-ingestion.
 * Useful for keeping the AvocadoDB index up-to-date during development.
 *
 * @example
 * ```typescript
 * import { FileMonitor } from 'avocadodb';
 *
 * const monitor = new FileMonitor({ intervalSeconds: 30 });
 * monitor.startMonitoring(['**\/*.ts', '**\/*.md']);
 *
 * // Monitor runs in background, auto-re-ingests changed files
 * // Stop when done: monitor.stopMonitoring();
 * ```
 *
 * @packageDocumentation
 */

import { spawn } from 'child_process';
import { existsSync, statSync } from 'fs';
import { homedir } from 'os';
import { join } from 'path';
import { glob } from 'glob';

/**
 * Callback for file change events
 */
export type OnChangeCallback = (files: string[]) => void;

/**
 * Background file watcher for automatic re-ingestion
 *
 * Monitors files matching specified patterns and automatically
 * re-ingests them when changes are detected. Runs with configurable
 * polling interval.
 *
 * @example
 * ```typescript
 * const monitor = new FileMonitor({ intervalSeconds: 30 });
 * monitor.startMonitoring(['docs/**\/*.md', 'src/**\/*.ts']);
 *
 * // Files are automatically re-ingested when modified
 * monitor.stopMonitoring();
 * ```
 */
export class FileMonitor {
  /** Polling interval in seconds */
  private readonly intervalSeconds: number;

  /** Path to ingest binary */
  private readonly ingestBinary: string | null;

  /** Monitor interval timer */
  private monitorInterval: NodeJS.Timeout | null = null;

  /** Last modified times for tracked files */
  private lastModified: Map<string, number> = new Map();

  /** Change event callback */
  private onChangeCallback: OnChangeCallback | null = null;

  /**
   * Create file monitor
   * @param options - Monitor options
   */
  constructor(options: {
    /** How often to check for changes in seconds (default: 30) */
    intervalSeconds?: number;

    /** Path to avocado ingest binary (auto-detected if not provided) */
    ingestBinary?: string;
  } = {}) {
    this.intervalSeconds = options.intervalSeconds || 30;
    this.ingestBinary = options.ingestBinary || this.findIngestBinary();
  }

  /**
   * Find avocado ingest binary
   * @returns Path to binary if found
   */
  private findIngestBinary(): string | null {
    const cwd = process.cwd();
    const home = homedir();

    const possiblePaths = [
      join(cwd, 'target/release/avocado'),
      join(cwd, '../target/release/avocado'),
      join(cwd, '../../target/release/avocado'),
      join(home, '.avocadodb/repo/target/release/avocado'),
      '/usr/local/bin/avocado',
    ];

    for (const path of possiblePaths) {
      if (existsSync(path)) {
        return path;
      }
    }

    return null;
  }

  /**
   * Register callback for file change events
   * @param callback - Function called with list of changed files
   *
   * @example
   * ```typescript
   * monitor.onChange((files) => {
   *   console.log(`Changed: ${files.join(', ')}`);
   * });
   * ```
   */
  onChange(callback: OnChangeCallback): void {
    this.onChangeCallback = callback;
  }

  /**
   * Monitor loop - checks for file changes
   */
  private async monitorLoop(patterns: string[], cwd: string): Promise<void> {
    if (!this.ingestBinary || !existsSync(this.ingestBinary)) {
      return;
    }

    try {
      // Find files matching patterns
      const pathsToCheck: string[] = [];
      for (const pattern of patterns) {
        try {
          const matching = await glob(pattern, {
            cwd,
            ignore: [
              '**/node_modules/**',
              '**/.git/**',
              '**/venv/**',
              '**/__pycache__/**',
              '**/target/**',
              '**/build/**',
              '**/dist/**',
              '**/.next/**',
              '**/.cache/**',
            ],
          });
          pathsToCheck.push(...matching.slice(0, 100));  // Limit per pattern
        } catch {
          // Ignore pattern errors
        }
      }

      // Check for modified files
      const changedFiles: string[] = [];
      for (const relPath of pathsToCheck) {
        const fullPath = join(cwd, relPath);
        if (!existsSync(fullPath)) continue;

        try {
          const stats = statSync(fullPath);
          if (!stats.isFile()) continue;

          const mtime = stats.mtimeMs;
          const lastMtime = this.lastModified.get(fullPath) || 0;

          // File is new or modified
          if (mtime > lastMtime) {
            // Re-ingest file
            await new Promise<void>((resolve) => {
              const proc = spawn(this.ingestBinary!, ['ingest', fullPath]);
              proc.on('close', (code) => {
                if (code === 0) {
                  this.lastModified.set(fullPath, mtime);
                  changedFiles.push(fullPath);
                }
                resolve();
              });
              proc.on('error', () => resolve());
            });
          }
        } catch {
          // Ignore file errors
        }
      }

      // Notify callback if files changed
      if (changedFiles.length > 0) {
        if (this.onChangeCallback) {
          this.onChangeCallback(changedFiles);
        } else {
          console.log(`🥑 Background: Re-ingested ${changedFiles.length} changed files`);
        }
      }
    } catch {
      // Silently continue on errors
    }
  }

  /**
   * Start background file monitoring
   * @param patterns - Glob patterns to monitor
   * @param cwd - Working directory to monitor (default: current directory)
   *
   * @example
   * ```typescript
   * monitor.startMonitoring([
   *   'docs/**\/*.md',
   *   'src/**\/*.ts',
   *   'README.md'
   * ]);
   * ```
   */
  startMonitoring(patterns: string[], cwd: string = process.cwd()): void {
    if (this.monitorInterval) {
      console.warn('⚠️  Monitor already running');
      return;
    }

    this.monitorInterval = setInterval(
      () => this.monitorLoop(patterns, cwd),
      this.intervalSeconds * 1000
    );
  }

  /**
   * Stop background file monitoring
   *
   * @example
   * ```typescript
   * monitor.stopMonitoring();
   * ```
   */
  stopMonitoring(): void {
    if (this.monitorInterval) {
      clearInterval(this.monitorInterval);
      this.monitorInterval = null;
    }
  }

  /**
   * Check if monitoring is active
   * @returns True if monitor is running
   */
  isMonitoring(): boolean {
    return this.monitorInterval !== null;
  }
}

File: sdks/python/avocado/monitor.py
"""Background File Monitoring for AvocadoDB.

Framework-agnostic file watcher that detects changes and triggers re-ingestion.
Useful for keeping the AvocadoDB index up-to-date during development.

Example:
    >>> from avocado import FileMonitor
    >>> monitor = FileMonitor(interval_seconds=30)
    >>> monitor.start_monitoring(["**/*.py", "**/*.md"])
    >>> # Monitor runs in background, auto-re-ingests changed files
"""

import subprocess
import threading
from pathlib import Path
from typing import Callable, Optional


class FileMonitor:
    """Background file watcher for automatic re-ingestion.

    Monitors files matching specified patterns and automatically
    re-ingests them when changes are detected. Runs in a background
    thread with configurable polling interval.

    Example:
        >>> from avocado import FileMonitor
        >>> monitor = FileMonitor(interval_seconds=30)
        >>> monitor.start_monitoring(["docs/**/*.md", "src/**/*.py"])
        >>> # Files are automatically re-ingested when modified
        >>> monitor.stop_monitoring()
    """

    def __init__(self, interval_seconds: int = 30, ingest_binary: Optional[Path] = None):
        """Initialize file monitor.

        Args:
            interval_seconds: How often to check for changes (default: 30)
            ingest_binary: Path to avocado ingest binary (auto-detected if None)
        """
        self.interval_seconds = interval_seconds
        self.ingest_binary = ingest_binary or self._find_ingest_binary()
        self._monitor_thread: Optional[threading.Thread] = None
        self._stop_flag = threading.Event()
        self._last_modified: dict[Path, float] = {}
        self._on_change_callback: Optional[Callable[[list[Path]], None]] = None

    def _find_ingest_binary(self) -> Optional[Path]:
        """Find avocado ingest binary.

        Returns:
            Path to ingest binary if found, None otherwise
        """
        possible_paths = [
            Path.cwd() / "target/release/avocado",
            Path.cwd().parent / "target/release/avocado",
            Path.cwd().parent.parent / "target/release/avocado",
            Path.home() / ".avocadodb/repo/target/release/avocado",
            Path("/usr/local/bin/avocado"),
        ]

        for path in possible_paths:
            if path.exists() and path.is_file():
                return path

        return None

    def on_change(self, callback: Callable[[list[Path]], None]):
        """Register callback for file change events.

        Args:
            callback: Function called with list of changed files

        Example:
            >>> def handle_changes(files):
            ...     print(f"Changed: {files}")
            >>> monitor.on_change(handle_changes)
        """
        self._on_change_callback = callback

    def _monitor_loop(self, patterns: list[str], cwd: Path):
        """Background monitoring loop.

        Args:
            patterns: Glob patterns to monitor
            cwd: Current working directory
        """
        while not self._stop_flag.is_set():
            try:
                # Wait for interval (allows early exit via stop_flag)
                if self._stop_flag.wait(timeout=self.interval_seconds):
                    break

                if not self.ingest_binary or not self.ingest_binary.exists():
                    continue

                # Find files matching patterns
                paths_to_check = []
                for pattern in patterns:
                    try:
                        matching = list(cwd.glob(pattern))
                        # Exclude common build/cache directories
                        matching = [
                            p for p in matching
                            if not any(part in p.parts for part in [
                                "node_modules", ".git", "venv", ".venv",
                                "__pycache__", "target", "build", "dist",
                                ".next", ".cache", ".tox"
                            ])
                        ]
                        paths_to_check.extend(matching[:100])  # Limit per pattern
                    except:
                        pass

                # Check for modified files
                changed_files = []
                for path in paths_to_check:
                    if not path.is_file():
                        continue

                    try:
                        mtime = path.stat().st_mtime
                        last_mtime = self._last_modified.get(path, 0)

                        # File is new or modified
                        if mtime > last_mtime:
                            # Re-ingest file
                            result = subprocess.run(
                                [str(self.ingest_binary), "ingest", str(path)],
                                capture_output=True,
                                timeout=10,
                            )
                            if result.returncode == 0:
                                self._last_modified[path] = mtime
                                changed_files.append(path)
                    except:
                        pass

                # Notify callback if files changed
                if changed_files:
                    if self._on_change_callback:
                        self._on_change_callback(changed_files)
                    else:
                        print(f"🥑 Background: Re-ingested {len(changed_files)} changed files")

            except Exception:
                pass  # Silently continue on errors

    def start_monitoring(self, patterns: list[str], cwd: Optional[Path] = None):
        """Start background file monitoring.

        Args:
            patterns: Glob patterns to monitor (e.g., ["**/*.py", "docs/**/*.md"])
            cwd: Working directory to monitor (default: current directory)

        Example:
            >>> monitor.start_monitoring([
            ...     "docs/**/*.md",
            ...     "src/**/*.py",
            ...     "README.md"
            ... ])
        """
        if self._monitor_thread and self._monitor_thread.is_alive():
            print("⚠️  Monitor already running")
            return

        cwd = cwd or Path.cwd()
        self._stop_flag.clear()
        self._monitor_thread = threading.Thread(
            target=self._monitor_loop,
            args=(patterns, cwd),
            daemon=True,
            name="AvocadoDB-FileMonitor"
        )
        self._monitor_thread.start()

    def stop_monitoring(self):
        """Stop background file monitoring.

        Example:
            >>> monitor.stop_monitoring()
        """
        if self._monitor_thread:
            self._stop_flag.set()
            self._monitor_thread.join(timeout=2)
            self._monitor_thread = None

    def is_monitoring(self) -> bool:
        """Check if monitoring is active.

        Returns:
            True if monitor thread is running, False otherwise
        """
        return self._monitor_thread is not None and self._monitor_thread.is_alive()


__all__ = ["FileMonitor"]
